Added jobs and projects pages for QFieldCloud (#197)
opengisch/QField-docs: documentation/reference/qfieldcloud/jobs.md

---
title: Jobs
---

Jobs on QFieldCloud perform heavy operation with project files and layers within QGIS. Jobs are created in response to certain user actions.

> Once a job is created, it enters a the project's job queue and waits for available QFieldCloud resources to run. At any given moment only one job can run per project. The rest of the queued jobs will perform in the same order as in they entered the queue.

Reviewer: "It enters a the" looks strange

Reviewer: ```suggestion
Once a job is created, it enters the project's job queue and waits for available QFieldCloud resources to run. At any given moment only one job can run per project. The rest of the queued jobs will be performed in the same order as they entered the queue.
```

Author: Shakespeare here

> Each job consists of one or more steps and each step is responsible for one self contained task on the project. Currently QFieldCloud supports only three predefined job workflows: `process_projectfile`, `package` and `delta_apply`.

Reviewer: ```suggestion
Each job consists of one or more steps and each step is responsible for one self contained task on the project. QFieldCloud supports three predefined job workflows: `process_projectfile`, `package` and `delta_apply`.
```

Author: Done

While running, jobs are writing log messages which are available on project's jobs page for jobs that have reached a final status `FINISHED` or `FAILED`.

Jobs have access to [project secrets](projects.md#secrets).

!!! note
    All jobs can be triggered using the QFieldCloud API.

!!! warning
    Any of the triggering conditions described on this page might change without notice.

!!! warning

>     All jobs must finished within 10 minutes or they will result in timeout error and will be terminated.

Reviewer: ```suggestion
    All jobs must finished within 10 minutes or they will result in a timeout error and will be terminated.
```

Author: "must finished" too :)

- There is already a `process_projectfile` job in `PENDING` status.

#### Troubleshoot

A `process_projectfile` job might result in `FAILED` status. Check the non-exhaustive list of causes bellow:

- The uploaded QGIS project file (`.qgs`/`.qgz`) is unreadable, incomplete, broken or wrong. Try to reupload the QGIS project file.
- QGIS is crashing after opening the project file. Try to identify the layer that is causing the crash by removing one layer at time from the project and reuploading the QGIS project file.

!!! NOTE
    Even if a `process_projectfile` job results in a `SUCCESS` status, it does not mean the project is properly configured. The `SUCCESS` status just states the project has been successfully opened and all the needed information has been extracted.

### Package (`package`) job

The `package` job convert a QGIS project to a QField project, the same way it is done on QGIS via QFieldSync. The `package` job will prepare all layers marked as "Offline editing" to a single GeoPackage.

#### Triggers

This job is triggered every time the **Download** or **Synchronize** buttons are pressed on QField. Unless at least one of the following condition are valid:

> - The project never run a `process_projectfile` job that resulted in `SUCCESS` status.

Reviewer: ```suggestion
- The project has never run a `process_projectfile` job that resulted in `SUCCESS` status.
```

- There is already a `package` job in `PENDING` status.

> - The project contains does not contain online vector layers (PostGIS, WFS etc), the latest `package` job result was `SUCCESS` and there were not file uploads, nor change uploads.

Reviewer: ```suggestion
- The project does not contain online vector layers (PostGIS, WFS etc), the latest `package` job result was `SUCCESS` and there were no file uploads, nor change uploads.
```

#### Troubleshoot

> A `package` job might result in `FAILED` status. Check the non-exhaustive list of causes bellow:

Reviewer: ```suggestion
A `package` job might result in `FAILED` status. Check the non-exhaustive list of causes below:
```

> - The project never run a `process_projectfile` job that resulted in `SUCCESS` status.

Reviewer: ```suggestion
- The project has never run a `process_projectfile` job that resulted in `SUCCESS` status.
```

> - Some of the project layers are inaccessible from QFieldCloud. Make sure all files are uploaded and all credentials to online layers (PostGIS, WFS etc) are stored within the QGIS project file.

Reviewer: For this section, would it be good to give a hint to the job log that could contain more information about the failure reason?

Author: I mentioned this in the begining and after that I have a section "Troubleshoot job logs" which should be enough. It would be great if we add a few screenshots here and there, will make the life of the reader easier too.

### Delta apply (`delta_apply`) job

Delta apply jobs is responsible to make all pushed QField changes permanent.

#### Triggers

> This job is triggered every time a **Synchronize** or **Push changes** buttons are pressed on QField, or **Apply pending changes** button is pressed on the **Changes** project page. If any of the following condition are valid:

Reviewer: ```suggestion
This job is triggered every time a **Synchronize** or **Push changes** button is pressed on QField, or **Apply pending changes** button is pressed on the **Changes** project page. If any of the following condition are valid:
```

- At least one of the online databases (PostGIS/WFS) used in the QGIS project reset the connection.
- The project is too big and the job has failed to run.
- There are hidden files and directories within the project that are preventing the normal work of QFieldCloud. Hidden files and directories are those starting with a leading dot (`.`).

## Troubleshoot job logs

When running a job, usually you can find a step called "Check project layers" that prints a table with all the project layers and status next to them.

The possible statuses are:

- **ok** - The layer loads correctly on QFieldCloud.
- **invalid_dataprovider** - the layer's data provider is invalid. Usually additional information is shown in the "Provider Summary".
- **invalid_layer** - This errors should happen very rarely if ever. The data is loaded correctly, but for some reason QGIS reports the layer as invalid.

### Unable to connect to service "`{SERVICE}`".

QFieldCloud tries to connect to a PostgreSQL service that is not available. You should need to create a new pgservice [secret](projects.md#secrets) so QFieldCloud can connect to the PostGIS service.

### Unable to connect to host "`{HOST}`".

> QFieldCloud cannot establish a connection to the given `{HOST}`. Your service is not accessible from the QFieldCloud server. You might been to ask your IT department to whitelist the QFieldCloud IP.

Reviewer: ```suggestion
QFieldCloud cannot establish a connection to the given `{HOST}`. Your service is not accessible from the QFieldCloud server. You might need to ask your IT department to whitelist the QFieldCloud IP.
```

Are we able to link to the qfc ip here?

Author: Good idea, however I prefer to not put it in the docs until we migrate to k8s and we fix the IP.

Reviewer: What do you think about a skeleton page that shows the state of affairs that can be adjusted later?

Author: If you mean an new page in the docs, then yes, I do agree. However, I believe it would be out of scope of the current PR.

Reviewer: A new page with technical specs would be great. Even if minimal with a "soon to be announced".

Author: See https://github.com/opengisch/QField-docs/pull/200
